### PR TITLE
[ENH] Spectogram wrapper

### LINKED ISSUE #1061: [ENH] Wrapping Spectrum generation of `scipy` as a series transformer
### Describe the feature or idea you want to propose

Some time series dataset require time-frequency analysis such as the spectrum. This spectrum can be generated through scipy

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.spectrogram.html

### Describe your proposed solution

It would be interesting if we wrap this transformation in a series transformer in `aeon`

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aadya940**
Closes #1061

We could also add a method to visualize the spectogram?

### CODE REVIEW

**Comment by aadya940** [comment]
Hi yes you're right, sorry for the trouble though. Will keep in mind :-)

**Comment by aadya940** [comment]
Everything seems to be working fine locally, why might the tests be failing?

**Comment by aadya940** [comment]
Everything seems to be working fine locally, why might the tests be failing?

**Comment by MatthewMiddlehurst** [comment]
I'll try to take a look tomorrow. Usually good to ask in Slack for feedback/reviews if testing is failing.

**Comment by MatthewMiddlehurst** [comment]
There appears to be an issue with the class docstring examples @aadya940 (`rng` does not exist) , see https://github.com/aeon-toolkit/aeon/actions/runs/8054369251/job/21998832063?pr=1237#step:7:266

These run in the test suite.

**Comment by MatthewMiddlehurst** [comment]
We have a `visualisation` module if you wish to create a function for plotting spectrograms, or you can create an issue.

**Comment by aadya940** [comment]
> There appears to be an issue with the class docstring examples @aadya940 (`rng` does not exist) , see https://github.com/aeon-toolkit/aeon/actions/runs/8054369251/job/21998832063?pr=1237#step:7:266
> 
> These run in the test suite.

This was hiding in plain sight ! Thank you very much :-)

File: aeon/visualisation/series/_series.py [hunk 1]
```diff
@@ -356,3 +358,69 @@ def plot_correlations(
...
+    ----------
+    sample_frequencies : ndarray
+                    Array of sample frequencies.
+    segment_time :   ndarray
+                Array of segment times.
+    spectrogram :     ndarray
+                Spectrogram of x. By default, the last axis of
+                spectrogram corresponds to the segment times.
+    return_both : boolean, default = False
+                Condition to allow negative frequency spectrum.
```
Comment (MatthewMiddlehurst) [inline aeon/visualisation/series/_series.py:382]: this is indented a bit weirdly

File: aeon/transformations/series/_spectrogram.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    >>> amp = 2 * np.sqrt(2)
+    >>> noise_power = 0.01 * fs / 2
+    >>> time = np.arange(N) / float(fs)
+    >>> mod = 500*np.cos(2*np.pi*0.25*time)
+    >>> carrier = amp * np.sin(2*np.pi*3e3*time + mod)
+    >>> noise = rng.normal(scale=np.sqrt(noise_power), size=time.shape)
+    >>> noise *= np.exp(-time/5)
+    >>> x = carrier + noise
+    >>> transformer = SpectrogramTransformer()  # doctest: +SKIP
+    >>> mp = transformer.fit_transform(x)  # doctest: +SKIP
```
Comment (MatthewMiddlehurst) [inline aeon/transformations/series/_spectrogram.py:43]: does this need to be skipped?

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
Thanks for the contribution. Left some comments. If possible could you add an example to the series plotting visualisation notebook in examples?

**Comment by MatthewMiddlehurst** [comment]
There are some minor pre-commit errors, you can either run it locally or i think the bot should fix them now that this is not a draft when the next commit is made.

**Comment by aadya940** [comment]
Thanks @MatthewMiddlehurst , Will do it ASAP! :-)

**Comment by aadya940** [comment]
Hi @MatthewMiddlehurst , I've done the necessary changes. However, I was thinking if we should by default return multiple values in `SpectrogramTransformer._transform` since they'll be used in plotting the frequency spectrum of the data.

**Comment by MatthewMiddlehurst** [comment]
It would make the default output of the transformer very inconsistent with other transformers if it had multiple outputs. If all three outputs are as important as the others, maybe a transformer is not the best way to go about this.

You may want to see the developer install instructions for setting up `pre-commit`, as there are failures which are usually automatically fixed by it. Maybe the bot does not work on forks.

**Comment by aadya940** [comment]
> It would make the default output of the transformer very inconsistent with other transformers if it had multiple outputs. If all three outputs are as important as the others, maybe a transformer is not the best way to go about this.
@MatthewMiddlehurst  Would it make sense to only implement the `plot_spectrogram` function for now which directly calls `scipy.signal.spectrogram` since plotting the frequency spectrum would be very useful as well?

**Comment by MatthewMiddlehurst** [comment]
You can do that if you want yeah. It just does not really fit the interface to return multiple things at once (ideally the output should always be a new series or feature vector, but thats less enforced as long as the output makes sense).

**Comment by aadya940** [comment]
@MatthewMiddlehurst Makes Sense 👍, I plan to implement standalone `plot_spectrogram` soon.

**Comment by MatthewMiddlehurst** [comment]
HI @aadya940 just checking up, need any help with this?

**Comment by aadya940** [comment]
Hey @MatthewMiddlehurst , This seems to be ready for review.

File: aeon/visualisation/series/_series.py [hunk 3]
```diff
@@ -356,3 +358,47 @@ def plot_correlations(
...
+
+    f, t, _spectrogram = spectrogram(series, fs=fs, return_onesided=return_onesided)
+    if not return_onesided:
+        ax.pcolormesh(t, fftshift(f), fftshift(_spectrogram, axes=0), shading="gouraud")
+        ax.set_ylabel("Frequency [Hz]")
+        ax.set_xlabel("Time [sec]")
+        return fig, ax
+
+    ax.pcolormesh(t, f, _spectrogram, shading="gouraud")
+    return fig, ax
```
Comment (MatthewMiddlehurst) [inline aeon/visualisation/series/_series.py:404]: It may be better to put this in an `else:` and have a single return for readability.

is the `return_onesided` plot fine without labels?

File: aeon/visualisation/series/tests/test_series_plotting.py [hunk 4]
```diff
@@ -316,3 +321,17 @@ def test_plot_correlations_arguments(series_to_plot, lags):
...
+    reason="skip test if required soft dependency not available",
+)
+def test_plot_spectrogram():
+    """Tests whether plot_spectrogram run with the given inputs or not."""
+    fig, ax = plot_spectrogram(y_airline, fs=1)
+    assert fig is not None
+    assert ax is not None
+    fig, ax = plot_spectrogram(y_airline, fs=1, return_onesided=False)
+    assert fig is not None
+    assert ax is not None
```
Comment (MatthewMiddlehurst) [inline aeon/visualisation/series/tests/test_series_plotting.py:337]: Could you try plotting in the test? See the above ones for setting the backend i.e. `matplotlib.use("Agg")`

**Review by MatthewMiddlehurst — COMMENTED** [review]
Thanks, looks good. Few comments.

Could you add a block to the `examples/visualisation` notebook so we can have a look at the output?

**Comment by aadya940** [comment]
Hey @MatthewMiddlehurst , did some changes! Thanks for reviewing :-)

**Comment by MatthewMiddlehurst** [comment]
Does the output look as you would expect? I'm not very familiar 🙂.

After testing is complete, I think only one plot output is required for the notebook. But that is optional, wont push for it.

**Comment by aadya940** [comment]
> Does the output look as you would expect? I'm not very familiar 🙂.
> 
> After testing is complete, I think only one plot output is required for the notebook. But that is optional, wont push for it.

I think this is good enough since we have all the internal frequencies of the time series visible in the plots. :-)  Also, I've added 2 plots in-case someone is confused about how to plot a two-sided spectrum.

**Review by MatthewMiddlehurst — APPROVED** [review]
